src/parser.rs

```diff
@@ -73,110 +75,182 @@ pub enum Fragment {
+            // FIXME: multi-character punctuation
+            TokenTree::Punct(punct) => Separator::Punct(punct),
+            TokenTree::Literal(literal) => Separator::Literal(literal),
+            TokenTree::Group(group) => {
+                return Err(Error::new(group.span(), "unexpected token"));
+            }
         })
-    ));
-}
-
-impl Synom for Matcher {
-    named!(parse -> Self, alt!(
-        do_parse!(
-            syn!(Dollar) >>
-            name: call!(Ident::parse_any) >>
-            punct!(:) >>
-            fragment: syn!(Fragment) >>
-            (Matcher::Fragment { name, fragment })
-        ) |
-        do_parse!(
-            syn!(Dollar) >>
-            content: parens!(many0!(syn!(Matcher))) >>
-            separator: option!(syn!(Separator)) >>
-            repetition: syn!(Repetition) >>
-            (Matcher::Repeat { content: content.1, separator, repetition })
-        ) |
-        delimited!(many0!(syn!(Matcher))) => {
-            |(delimiter, content)| Matcher::Group { delimiter, content }
-        } |
-        do_parse!(
-            not!(syn!(Dollar)) >>
-            punct: syn!(Punct) >>
-            (Matcher::Punct(punct))
-        ) |
-        call!(Ident::parse_any) => { Matcher::Ident } |
-        syn!(Literal) => { Matcher::Literal }
-    ));
-}
-
-impl Synom for Repetition {
-    named!(parse -> Self, alt!(
-        punct!(*) => { |_| Repetition::Repeated } |
-        punct!(+) => { |_| Repetition::AtLeastOnce } |
-        punct!(?) => { |_| Repetition::AtMostOnce }
-    ));
-}
-
-impl Synom for Separator {
-    named!(parse -> Self, alt!(
-        do_parse!(
-            not!(syn!(Repetition)) >>
-            punct: syn!(Punct) >>
-            (Separator::Punct(punct))
-        ) |
-        call!(Ident::parse_any) => { Separator::Ident } |
-        syn!(Literal) => { Separator::Literal }
-    ));
-}
-
-impl Synom for Fragment {
-    named!(parse -> Self, alt!(
-        custom_keyword!(ident) => { |_| Fragment::Ident } |
-        custom_keyword!(path) => { |_| Fragment::Path } |
-        custom_keyword!(expr) => { |_| Fragment::Expr } |
-        custom_keyword!(ty) => { |_| Fragment::Ty } |
-        custom_keyword!(pat) => { |_| Fragment::Pat } |
-        custom_keyword!(stmt) => { |_| Fragment::Stmt } |
-        custom_keyword!(block) => { |_| Fragment::Block } |
-        custom_keyword!(item) => { |_| Fragment::Item } |
-        custom_keyword!(meta) => { |_| Fragment::Meta } |
-        custom_keyword!(tt) => { |_| Fragment::Tt } |
-        custom_keyword!(vis) => { |_| Fragment::Vis } |
-        custom_keyword!(literal) => { |_| Fragment::Literal } |
-        custom_keyword!(lifetime) => { |_| Fragment::Lifetime }
-    ));
-}
-
-pub fn parse(src: &str) -> Result<MacroRules, syn::synom::ParseError> {
+    }
+}
+
+impl Parse for Repetition {
+    fn parse(input: ParseStream) -> Result<Self> {
+        if input.parse::<Option<Token![*]>>()?.is_some() {
```

Comment:
My spider senses are tickling here: IIRC there is a change in the 2018 edition in order to disambiguate repetition markers from separators. Maybe we will need a knob here in the future?